### PR TITLE
Update Manual

### PR DESCRIPTION
**Pull request by naqvisat**
Update manual and changes related to pattern.

### CODE REVIEW

File: manual/introduction.md [hunk 1]
```diff
@@ -10,14 +10,30 @@ libraryDependencies += "pl.metastack" %%  "metatime" % "%version%"  // Scala
 libraryDependencies += "pl.metastack" %%% "metatime" % "%version%"  // Scala.js
 ```
 
-## Example
-[scala type="section" value="introduction" file="Examples"]
+## Key Features
```
Comment (tindzk) [inline manual/introduction.md:13]: I would suggest to keep the features list and create a chapter for each of them. The chapters need to be registered in `Manual` (line 51).

File: manual/introduction.md [hunk 2]
```diff
@@ -10,14 +10,30 @@ libraryDependencies += "pl.metastack" %%  "metatime" % "%version%"  // Scala
...
+## Type-safe DSL
+  MetaTime is a type-safe DSL which offers to prevents a lot of type errors a user might do and gives users opportunity to focus
```
Comment (tindzk) [inline manual/introduction.md]: Could you rephrase this and give a couple of examples? For example, involving addition, `fromNow` etc.

File: manual/src/main/scala/pl/metastack/metatime/manual/Manual.scala [hunk 3]
```diff
@@ -121,7 +121,7 @@ object Manual extends App with Shared {
     Paths.get(from) -> Paths.get(to)
   }.foreach { case (from, to) =>
     if (Files.exists(from)) Files.delete(from)
-    Files.createSymbolicLink(from, to)
+    Files.createSymbolicLink(from, to) //TODO
```
Comment (tindzk) [inline manual/src/main/scala/pl/metastack/metatime/manual/Manual.scala]: Remove comment

File: shared/src/main/scala/pl/metastack/metatime/Pattern.scala [hunk 4]
```diff
@@ -1,6 +1,6 @@
 package pl.metastack.metatime
 
-case class Pattern(parts: Seq[DateTime => String] = Nil) {
+case class Pattern(parts: Seq[Component => String] = Nil) {
```
Comment (tindzk) [inline shared/src/main/scala/pl/metastack/metatime/Pattern.scala]: Can you change this back to `DateTime`?

File: shared/src/main/scala/pl/metastack/metatime/Pattern.scala [hunk 5]
```diff
@@ -69,5 +91,16 @@ object Pattern {
   val HoursAgo  : Pattern = ago(Pattern().minute.space.string("hour(s)"))
   val DaysAgo   : Pattern = ago(Pattern().minute.space.string("day(s)"))
 
-  def branch(ranges: Seq[(Range, Pattern)], fallback: Pattern): Pattern = ???
+  def branch(ranges: Seq[(Range, Pattern)], fallback: Pattern): Pattern = {
```
Comment (tindzk) [inline shared/src/main/scala/pl/metastack/metatime/Pattern.scala]: As discussed, you have to write something along the lines of:

``` scala
val pattern = Pattern()
pattern.add { time =>
  // Find range that matches `time`
  ...
}
pattern
```

File: shared/src/main/scala/pl/metastack/metatime/Range.scala [hunk 6]
```diff
@@ -1,5 +1,4 @@
 package pl.metastack.metatime
 
-trait Range {
-
+case class Range (lowerBound: Long, upperBound: Long) {
```
Comment (tindzk) [inline shared/src/main/scala/pl/metastack/metatime/Range.scala]: You can remove the curly braces. Also remove the space before the left parenthesis.

File: shared/src/main/scala/pl/metastack/metatime/Pattern.scala (hunk 4 above)
Comment (naqvisat) [inline shared/src/main/scala/pl/metastack/metatime/Pattern.scala]: Actually if I change it back to DateTime then I cant invoke format function for Time instance for following testcase:
    `val time = Time(11, 1, 0, 0)`
    `assert(time.format(Pattern.DefaultTime) == "11:01 AM")`

Comment (tindzk) [inline shared/src/main/scala/pl/metastack/metatime/Pattern.scala]: As you can see, some methods in `Pattern` are well-defined only for `DateTime` objects, while others expect `Time` objects. I would suggest to create add a type parameter to `Pattern` and make it a trait instead. Then, create two subclasses `DateTimePattern` and `TimePattern` which parameterise `Pattern` accordingly.

File: shared/src/main/scala/pl/metastack/metatime/PatternDateTime.scala [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+  def monthName(short: Boolean = false): PatternDateTime = {
+    add(dateTime =>
+      months(dateTime.month - 1))
+  }
+
+  def day: PatternDateTime = add(_.day.toInt.toString)
+
+  def amPm: PatternDateTime =
+    add(dt =>
+      if (dt.asInstanceOf[Time].h > 12) "PM"
```
Comment (tindzk) [inline shared/src/main/scala/pl/metastack/metatime/PatternDateTime.scala]: Is the cast needed?

File: shared/src/main/scala/pl/metastack/metatime/PatternTime.scala [hunk 8]
```diff
@@ -0,0 +1,66 @@
+package pl.metastack.metatime
+
+case class PatternTime(parts: Seq[Time => String] = Nil) {
+  def string(s: String): PatternTime = add(_ => s)
+
+  def dash : PatternTime = string("-")
```
Comment (tindzk) [inline shared/src/main/scala/pl/metastack/metatime/PatternTime.scala]: This is duplicated in `PatternDateTime`. Can you create a trait `Pattern` that contains shared functionality?

File: shared/src/main/scala/pl/metastack/metatime/Scheduler.scala [hunk 9]
```diff
@@ -41,6 +41,10 @@ trait Scheduler {
     schedule(dateTime.milliseconds.millis)(f)
   def every(time: Offset[Component])(f: => Unit): Cancelable =
     schedule(Math.abs(time.component.milliseconds()).millis)(f)
+
+  override def toString: String = {
```
Comment (tindzk) [inline shared/src/main/scala/pl/metastack/metatime/Scheduler.scala]: Can you remove this function? A scheduler is not a task.

File: shared/src/test/scala/pl/metastack/metatime/SchedulerTest.scala [hunk 10]
```diff
@@ -41,8 +40,7 @@ class SchedulerTest extends AsyncFunSuite {
     val timeBefore = Time.now()
     val p = Promise[Boolean]()
     scheduler.every(Time(0, 0, 1, 0)) {
-      // TODO Check whether every() is called more than once in the same intervals
```
Comment (tindzk) [inline shared/src/test/scala/pl/metastack/metatime/SchedulerTest.scala:44]: Please keep this comment because we don't check yet whether `every()` is correctly implemented.

File: shared/src/main/scala/pl/metastack/metatime/PatternTime.scala (hunk 8 above)
Comment (naqvisat) [inline shared/src/main/scala/pl/metastack/metatime/PatternTime.scala]: Actually I tried to create Pattern to have common functions and i put following functions there:

  /_def dash : PatternDateTime = string("-")
  def space: PatternDateTime = string(" ")
  def colon: PatternDateTime = string(":")
  def comma: PatternDateTime = string(",")
  def dot  : PatternDateTime = string(".")_/

But this requires me to defined string function also so I did

 def string(s: String): PatternDateTime = add(_ => s)

but then it requires me to define add method as well while we have add methods specific for DateTime and Time (separately). What do you suggest?

Comment (tindzk) [inline shared/src/main/scala/pl/metastack/metatime/PatternTime.scala]: You can add a type parameter (`T <: Pattern`) to `Pattern`. In `PatternDateTime` extend from `Pattern[PatternDateTime]`. Then, make the method `add` abstract and return `T`. You will need a second type parameter which you initialise with `DateTime`.

Comment (naqvisat) [inline shared/src/main/scala/pl/metastack/metatime/PatternTime.scala]: Can you please have a look at this gist: https://gist.github.com/naqvisat/55f7cf0e21f39b1b2b62e0cde61d4826

**Comment by tindzk** [comment]
Thanks!
